[5기 이지영]자바스크립트 계산기미션

gygy7151 opened the pull request.

gygy7151 committed:
21.07.13 계산기미션

titiman1013 commented on calculator.html:
```diff
@@ -0,0 +1,117 @@
+    
+    function calculate() {
+      var display = document.getElementById('display');
+      var result = eval(display.value);
```
eval함수는 단점이 많아 사용하지 않는게 좋다고 합니다 !

titiman1013 commented on calculator.html:
```diff
@@ -0,0 +1,117 @@
+
+    }
+
+    //초기화 함수명은 clear를 사용할수없다 js내장함수명중 clear 존재하기때문.
```
배워갑니다

titiman1013 reviewed (commented):
안녕하세요 이지영님
코드리뷰를 함께하게된 노현석입니다.

기능이 모두 완성되지 않아 조금 아쉬웠지만, 
제가 js언어에 익숙하지 않은데도 기능별 함수처리로 가독성이 좋았던것 같습니다.

EungyuCho commented on calculator.html:
```diff
@@ -0,0 +1,117 @@
+      .a3{
+        background-color: orange;
+        color: white;
+      }
```
```suggestion
      .bg-grey {
        background-color: gray;
      }
      
      .font-white {
        color: white;
      }
      
      .bg-smoke{
        background-color: whitesmoke;
      }
      
      .bg-orange {
        background-color: orange;
      }
```

디자인 라이브러리처럼 재활용성이 좋네요!

이런 식으로 작성하면 다시 사용할 때 어떤 역할을 하는지 더 파악하기 쉬워질 것 같아요 😄

EungyuCho commented on calculator.html:
```diff
@@ -0,0 +1,117 @@
+      <td class="a2" onclick="add(0)" colspan="2">0</td>
+      <td class="a2" onclick="add('.')">.</td>
+      <td class="a3" onclick="calculate()">=</td>
+    </tr>
```
다음번에 `이벤트 위임` 키워드에 대해서 공부해보는 것도 좋을 것 같아요!

이벤트 리스너를 여러 개 등록하다 보면 메모리 누수가 있을 수 있어서 위임을 통해서 등록하는 방법을 추천드립니다!!

[이 글](https://ui.toast.com/weekly-pick/ko_20160826)이 도움이 되실 것 같아요 👍

jparkk0517 commented on calculator.html:
```diff
@@ -0,0 +1,117 @@
+        if(isNan(char) ==true) {
+
+        } else { //연산자가 아니라면
+          document.getElementById('display').value += char;
```
document.getElementById 를 통한 DOM객체 접근 함수가 반복되고있는것 같아요. 하나의 변수에 할당해 놓고 이용하시는 편이 더 효율적일것 같아요!

EungyuCho commented on calculator.html:
```diff
@@ -0,0 +1,117 @@
+  </table>
+  <script>
+    //식이아닌 문자열입력방지장치 : 이전눌렀던 값이 숫자이면 true, 아니면 false;
+    var numberClicked = false; //전역변수
```
`var`보다 최근에는 대부분 `const`나 `let`을 통한 변수 선언을 활용하고 있어요!

`var`로 작성할 시 중복 선언이 가능하기 때문에 예상치 못한 동작에 대한 리스크도 있다고 생각이 되어서 `const`와 `let`에 친숙해지는 걸 추천드리고 싶어요.

[참고](https://curryyou.tistory.com/192)

jparkk0517 replied:
이 부분 아이디어가 굉장히 좋은것 같아요. digit과 operation의 입력을 구분하도록 하신거요.

jparkk0517 commented on calculator.html:
```diff
@@ -0,0 +1,117 @@
+    function add(char) {
+
+          //숫자나 연산자 버튼을 누르면, add함수가 호출되고, 사용자가 누른값이 char에 들어감
+      if (numberClicked == false) {
```
numberClicked 자체가 boolean형이기 때문에 비교연산자를 넣지 않아도 될것같아요. 만약 가독성 측면에서 true false를 알고자 하신다면 변수명을 변경하셔서 넣으시는 방법은 어떨까요?

EungyuCho replied:
```suggestion
      if (!numberClicked) {
```
```suggestion
      if (numberClicked === false) {
```
`==`의 경우 
저는 코드를 작성할 때 자바스크립트의 경우 조건문 판단에 `==`(동등 연산자)보다 정확한 판단을 위해서 `===`(일치) 연산자를 통해서 판단하는 경우가 많더라구요. `==`의 경우 값이 동등한지만 비교하기 때문에 타입을 강제적으로 변경해서 타입이 다르더라도 값만 같다면 조건에 맞다고 판단할 수 있어요.

엄격하게 비교하는 `===`를 추천드리고 맨 위의 제안처럼 변수 앞에 느낌표를 통해서 작성할 수도 있어요!

jparkk0517 commented on calculator.html:
```diff
@@ -0,0 +1,117 @@
+
+          //숫자나 연산자 버튼을 누르면, add함수가 호출되고, 사용자가 누른값이 char에 들어감
+      if (numberClicked == false) {
+        if(isNan(char) ==true) {
```
isNaN이 오타가 나신걸까요?

EungyuCho replied:
```suggestion
        if(isNaN(char)) {
```
마지막 대문자 N이에요!

jparkk0517 commented on calculator.html:
```diff
@@ -0,0 +1,117 @@
+      <td class="a3" onclick="add('/')">%</td>
+    </tr>
+    <tr>
+      <td class="a2" onclick="add(7)">7</td>
+      <td class="a2" onclick="add(8)">8</td>
+      <td class="a2" onclick="add(9)">9</td>
+      <td class="a3" onclick="add('*')">*</td>
+    </tr>
+    <tr>
+      <td class="a2" onclick="add(4)">4</td>
+      <td class="a2" onclick="add(5)">5</td>
+      <td class="a2" onclick="add(6)">6</td>
+      <td class="a3" onclick="add('-')">-</td>
+    </tr>
+    <tr>
+      <td class="a2" onclick="add(1)">1</td>
+      <td class="a2" onclick="add(2)">2</td>
+      <td class="a2" onclick="add(3)">3</td>
+      <td class="a3" onclick="add('+')">+</td>
+    </tr>
+    <tr>
+      <td class="a2" onclick="add(0)" colspan="2">0</td>
+      <td class="a2" onclick="add('.')">.</td>
+      <td class="a3" onclick="calculate()">=</td>
```
인라인 태그를 이용하여 onClick이벤트를 주셔서 스크립트가 크게 줄어들었군요!!

스크립트가 확 줄어드는 장점을 취하신 대신 
소스가 마크업 랭귀지와 스크립트 언어로 분포되는 단점, 정적인 HTML에 하드코딩을 해야 하는 단점을 취하신 것 같아요.

확장성을 고려하신다면 스크립트를 통한 DOM 객체 핸들링이 조금 더 효율적이지 않을까요?

EungyuCho reviewed (commented):
안녕하세요! 5기 스터디 참가자 조은규입니다.

아직 완성을 못하셔서 조금 아쉽지만 노력하신 점이 잘 보여서 좋네요 👍 
바꿔보면 좋을 것 같은 점들을 리뷰해봤어요 참고해 주시면 좋을 것 같아요!

아직 시간은 많으니까 너무 급하지 않게 공부하시는 게 좋을 것 같아요.
너무 급하게 달리다 보면 번아웃이 크게 오기도 하고 다시 일어나는데 시간이 많이 걸리더라고요..
방향성만 잘 맞춘다면 노력한 분들은 다 잘 되시더라구요.

조급하지 않되 꾸준히!... 지영님 블랙커피 스터디 완주까지 응원할게요! 파이팅입니다!!! 💯

jparkk0517 reviewed (commented):
이지영님 안녕하세요^^

코드리뷰를 하게 된 어호선입니다.

flag값을 통해 digit과 operation을 분리하신 부분이 매우 참신했습니다.

기능 구현사항에 포함된 소수점 절삭, 숫자 3자리 이상 입력 불가, alert창 표시 등의 기능들이 구현되지 않은것 같아요.

HTML 태그에 인라인으로 onClick이벤트를 바인딩하는것은 컴포넌트 형식으로 만들어지는 react, vue와 같은 라이브러리들에서 많이 이용되더라구요. 그러나 컴포넌트 단위로 파일이 나뉘어지고, Function / Class로 구현해놓아 가독성을 높여주는 react, vue와는 다르게 vanilla.js를 사용할 경우 인라인 태그를 통해 이벤트를 바인딩하게 되면 소스가 여러곳에 산재하게 되어 내용을 찾을때 번거로운 단점이 있는것 같습니다.

gygy7151 committed:
21.07.24_cypress환경설정

gygy7151 committed:
210724_cypress기본환경설정최종

#제목과 상동

gygy7151 committed:
210725_최종수정본

#3자리 숫자이상 제한기능 구현못함
#계산 결과를 표현할 때 소수점 이하는 버림하는 기능도 구현못함
나머지 기능에 대한 테스트 코드는 작성완료.
하지만 Cypress 통한 E2E테스트코드는 못짬
